### PR TITLE
storage: correctly transition the envelope_state_bytes value when moving moving from memory to rocksdb

### PR DESCRIPTION
**Pull request by guswynn**
We noticed during an investigation that `mz_envelope_state_bytes` seemed to _double-count_ the actual size of the 

This is because we have this bug, which erroneously always subtracts 0 when transitioning this size from the memory-size to the rocksdb size. This was not noticed in the past because:

- `workflow_autospill` was not actually testing the right codepath
  - even if it was, I think using more than 1 worker can mask the issue
- before correctness property 2 was fixed, the size problem was only for a single buffer's work of kafka messages. Now, during hydration, the entire snapshot is double-counted.
- Its fixed after the first rehydration.

This pr fixes all these issues AND adds an additional commit to ensure the test actually is testing what its supposed to be testing. 

### Motivation
  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/render/upsert/types.rs [hunk 1]
```diff
@@ -448,8 +448,6 @@ pub struct InMemoryHashMap {
 
 impl InMemoryHashMap {
     fn drain(&mut self) -> Drain<'_, UpsertKey, StateValue> {
-        // clearing the total_size when draining
-        self.total_size = 0;
```
Comment (rjobanp) [inline src/storage/src/render/upsert/types.rs:452]: I'm not following why this isn't correct? The `self.state.drain()` will clear the hashmap and we want this value to match the resulting size of the hashmap, right?

If the caller of `drain()` needs access to the size of the hashmap before they call `drain`, maybe we should return it too `(iterator_size, self.state.drain())`?

Comment (rjobanp) [reply src/storage/src/render/upsert/types.rs:452]: E.g. here https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/render/upsert/types.rs#L622 we have access to `in_memory_size` and if that's guaranteed to not change then we could use `put_stats.size_diff -= in_memory_size;`?

Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:452]: updated!

**Review by rjobanp — APPROVED** [review]

**Comment by moulimukherjee** [comment]
Oof, thanks for fixing this 🙏

**Pull request merged**